oak_attestation_verification/src/verifier.rs

```diff
@@ -121,43 +122,166 @@ fn verify_dice_chain(evidence: &Evidence) -> anyhow::Result<()> {
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no root layer reference values"))?;
+
+    // Verify Transparent Release endorsements in root layer.
+    verify_root_layer_endorsements(now_utc_millis, e0, r0)?;
+
+    // Verify Transparent Release endorsements in kernel layer.
+    verify_kernel_layer_endorsements(now_utc_millis, e1, r1)?;
+
+    // Verify Transparent Release endorsements in application layer.
+    let e20 = e2
+        .binary
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no binary endorsement"))?;
+
+    let r20 = r2
+        .binary
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no binary reference value"))?;
+
+    verify_transparent_release_endorsement(now_utc_millis, e20, r20)?;
+
+    let e21 = e2
+        .configuration
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no configuration endorsement"))?;
+
+    let r21 = r2
+        .configuration
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no configuration reference value"))?;
+
+    verify_transparent_release_endorsement(now_utc_millis, e21, r21)
 }
 
 fn verify_oak_containers(
     now_utc_millis: i64,
-    _evidence: &Evidence,
     endorsements: &OakContainersEndorsements,
     reference_values: &OakContainersReferenceValues,
 ) -> anyhow::Result<()> {
-    let ends_layer = endorsements
+    // Bind references to layers to variables.
+    let e0 = endorsements
         .root_layer
         .as_ref()
         .ok_or_else(|| anyhow::anyhow!("no root layer endorsements"))?;
+    let e1 = endorsements
+        .kernel_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no kernel layer endorsements"))?;
+    let e2 = endorsements
+        .system_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no system layer endorsements"))?;
+    let e3 = endorsements
+        .container_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no container layer endorsements"))?;
 
-    let ref_layer = reference_values
+    let r0 = reference_values
         .root_layer
         .as_ref()
         .ok_or_else(|| anyhow::anyhow!("no root layer reference values"))?;
+    let r1 = reference_values
+        .kernel_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no root layer reference values"))?;
+    let r2 = reference_values
+        .system_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no root layer reference values"))?;
+    let r3 = reference_values
+        .container_layer
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("no root layer reference values"))?;
 
-    let stage0_end = ends_layer
-        .stage0
+    // Verify Transparent Release endorsements in root layer.
+    verify_root_layer_endorsements(now_utc_millis, e0, r0)?;
+
+    // Verify Transparent Release endorsements in kernel layer.
+    verify_kernel_layer_endorsements(now_utc_millis, e1, r1)?;
+
+    // Verify Transparent Release endorsements in system layer.
```

Comment:
Should this be extracted into a separate verify_system_layer_endorsements? In general, having a 1:1 correspondence between separate proto messages and these functions will probably make maintenance easier